collectors/channels_collector.go

```diff
@@ -317,6 +340,150 @@ func (c *ChannelsCollector) Collect(ch chan<- prometheus.Metric) {
+}
+
+// getInboundFee calculates to forward fee for a specific amount charged by the
+// last hop before this node.
+func getInboundFee(amt btcutil.Amount, remotePolicies map[uint64]*lnrpc.RoutingPolicy,
+	remoteBalances map[uint64]btcutil.Amount) *btcutil.Amount {
+
+	var fee btcutil.Amount
+
+	// Copy the remote balances so they can be decreased as we find shards.
+	remainingBalances := make(map[uint64]btcutil.Amount)
+	for ch, balance := range remoteBalances {
+		remainingBalances[ch] = balance
+	}
+
+	// Assume a perfect mpp splitting algorithm that knows exactly how much
+	// can be sent through each channel. This is a simplification, because
+	// in reality senders need to trial and error to find a shard amount
+	// that works.
+	amountRemaining := amt
+	for amountRemaining > 0 {
+		var (
+			bestChan        uint64
+			bestSpecificFee float64
+			bestAmount      btcutil.Amount
+			bestFee         btcutil.Amount
+		)
+
+		// Find the best channel to send the amount or a part of the
+		// amount.
+		for ch, balance := range remainingBalances {
+			// Skip channels without remote balance.
+			if balance == 0 {
+				continue
+			}
+
+			policy, ok := remotePolicies[ch]
+			if !ok {
+				continue
+			}
+
+			// Cap at the maximum receive amount for this channel.
+			amountToSend := amountRemaining
+			if amountToSend > balance {
+				amountToSend = balance
+			}
+
+			// Calculate fee for this amount to send.
+			fee := btcutil.Amount(
+				policy.FeeBaseMsat/1000 +
+					int64(amountToSend)*policy.FeeRateMilliMsat/1000000,
+			)
+
+			// Calculate the specific fee for this amount, being the
+			// fee per sat sent.
+			specificFee := float64(fee) / float64(amountToSend)
+
+			// Select the best channel for this shard based on the
+			// lowest specific fee.
+			if bestChan == 0 || bestSpecificFee > specificFee {
+				bestChan = ch
+				bestSpecificFee = specificFee
+				bestAmount = amountToSend
+				bestFee = fee
+			}
+		}
+
+		// No liquidity to send the full amount, break.
+		if bestChan == 0 {
+			return nil
+		}
+
+		amountRemaining -= bestAmount
+		fee += bestFee
+		remainingBalances[bestChan] -= bestAmount
+	}
+
+	return &fee
+}
+
+// getRemotePolicies gets all the remote policies for enabled channels of this
+// node's peers.
+func (c *ChannelsCollector) getRemotePolicies(pubkey string) (
+	map[uint64]*lnrpc.RoutingPolicy, error) {
+
+	nodeInfoResp, err := c.lnd.GetNodeInfo(
+		context.Background(), &lnrpc.NodeInfoRequest{
+			IncludeChannels: true,
+			PubKey:          pubkey,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	policies := make(map[uint64]*lnrpc.RoutingPolicy)
+	for _, i := range nodeInfoResp.Channels {
+		var policy *lnrpc.RoutingPolicy
+		switch {
+		case i.Node1Pub == pubkey:
+			if i.Node2Policy != nil {
+				policy = i.Node2Policy
+			}
+
+		case i.Node2Pub == pubkey:
+			if i.Node1Policy != nil {
+				policy = i.Node1Policy
+			}
+
+		default:
+			return nil, fmt.Errorf("pubkey not in node info channels")
```

Comment:
Assuming this happens, what if we simply log and continue?

Reply:
Yes, that is always the discussion. How hard to error out. I did a return because it should never happen.

Reply:
Yeah maybe doing that is better but also not sure if the metric will approximate the missing value or will default to zero here?

Reply:
It will not show a value at all. Missing.

Reply:
sgtm